Detectors/CTP/reconstruction/include/CTPReconstruction/CTFCoder.h

```diff
@@ -151,27 +163,118 @@ o2::ctf::CTFIOSize CTFCoder::decode(const CTF::base& ec, VTRG& data, LumiInfo& l
+          // digitsMap[irs].printStream(std::cout);
+        }
+      }
+    } else { // correction would make IR prior to mFirstTFOrbit, skip
+      // itInp += CTFHelper::CTPInpNBytes;
       itCls += CTFHelper::CTPClsNBytes;
-      continue;
+      // continue;
     }
-    auto& dig = data.emplace_back();
-    dig.intRecord = ir - mBCShift;
-    for (int i = 0; i < CTFHelper::CTPInpNBytes; i++) {
-      dig.CTPInputMask |= static_cast<uint64_t>(*itInp++) << (8 * i);
+  }
+  if (mDecodeInps) {
+    std::vector<CTPDigit> digits;
+    o2::ctp::RawDataDecoder::shiftInputs(digitsMap, digits, mFirstTFOrbit);
+    for (auto const& dig : digits) {
+      data.emplace_back(dig);
     }
-    for (int i = 0; i < CTFHelper::CTPClsNBytes; i++) {
-      dig.CTPClassMask |= static_cast<uint64_t>(*itCls++) << (8 * i);
+  } else {
+    for (auto const& dig : digitsMap) {
+      data.emplace_back(dig.second);
     }
   }
   assert(itInp == bytesInput.end());
   assert(itCls == bytesClass.end());
   iosize.rawIn = header.nTriggers * sizeof(CTPDigit);
   return iosize;
 }
+///________________________________
+template <typename CTF>
+bool CTFCoder::finaliseCCDB(o2::framework::ConcreteDataMatcher& matcher, void* obj)
+{
+  LOG(info) << "CTFCoder::finaliseCCDB called";
+  bool match = false;
+  if (mLoadDictFromCCDB && (match = (matcher == o2::framework::ConcreteDataMatcher(mDet.getDataOrigin(), "CTFDICT", 0)))) {
+    const auto* dict = (std::vector<char>*)obj;
+    if (dict->empty()) {
+      LOGP(info, "Empty dictionary object fetched from CCDB, internal per-TF CTF Dict will be created");
+    } else {
+      std::vector<char> bufVec;
+      if (isTreeDictionary(obj)) {
+        auto v = loadDictionaryFromTree<CTF>(reinterpret_cast<TTree*>(obj));
+        bufVec.swap(v);
+        dict = &bufVec;
+      }
+      createCoders(*dict, mOpType);
+      mExtHeader = static_cast<const o2::ctf::CTFDictHeader&>(CTF::get(dict->data())->getHeader());
+      LOGP(info, "Loaded {} from CCDB", mExtHeader.asString());
+    }
+    mLoadDictFromCCDB = false; // we read the dictionary at most once!
+  } else if ((match = (matcher == o2::framework::ConcreteDataMatcher("CTP", "Trig_Offset", 0)))) {
+    const auto& trigOffsParam = o2::ctp::TriggerOffsetsParam::Instance();
+    auto bcshift = trigOffsParam.customOffset[mDet.getID()];
+    auto bcshiftInp = trigOffsParam.globalInputsShift;
+    if (bcshift || bcshiftInp) {
+      if (mSupportBCShifts) {
+        LOGP(info, "Decoded IRs will be augmented by {} BCs, discarded if become prior to 1st orbit", bcshift);
+        setBCShift(-bcshift); // offset is subtracted
+        setBCShiftInputs(-bcshiftInp);
+        LOG(info) << "BC shifts: Trig Classes:" << bcshift << " Inputs:" << bcshiftInp;
+      } else {
+        LOGP(alarm, "Decoding with {} BCs shift is requested, but the {} does not support this operation, ignoring request", bcshift, mDet.getName());
+      }
+    }
+  }
+  return match;
+}
```

Comment:
This is not needed, all needed objects are already loaded by the base CTFCoderBase, you just need to assign the new shift. So, use 
{
  CTFCoderBase::finaliseCCDB<CTF>(matcher, obj);
  setBCShiftInputs(-trigOffsParam.customOffset[mDet.getID()];);
}

Reply:
Ruben I am not sure we understand each other.:
in both ctp decoder and ctf decoder I want to have two independent global shifts:
1.) trigger class mask gloabl shift trigOffsParam.customOffset[mDet.getID() as before ( because it was used for triggerclass only)
2.) input class mask global shift trigOffsetParam.globalInputsShift

Reply:
The `CTFCoderBase::finaliseCCDB` base class method already does all what you have in this piece of code, except calling setting the BCShiftInputs. So, all what you need to do is to call the base method then set the shift (in the prev. version there was a typo, I mentioned the existing shift), i.e.
```
template <typename CTF=o2::ctp::CTF>
void CTFCoderBase::assignDictVersion(CTFDictHeader& h) const
{
  CTFCoderBase::finaliseCCDB<CTF>(matcher, obj);
  setBCShiftInputs(-o2::ctp::TriggerOffsetsParam::Instance().globalInputsShift);
}
```

Reply:
The globalInputShift will be loaded with all other offsets from CCDB and can be overridden from the command line.
The CCDB request is here: https://github.com/AliceO2Group/AliceO2/blob/a4932790654058445013ecd3dabc80b6972b8fdd/Detectors/CTP/workflow/src/EntropyDecoderSpec.cxx#L87

Reading is done here: https://github.com/AliceO2Group/AliceO2/blob/a4932790654058445013ecd3dabc80b6972b8fdd/Detectors/Base/src/CTFCoderBase.cxx#L54
and finalized where you copied your code from: https://github.com/AliceO2Group/AliceO2/blob/a4932790654058445013ecd3dabc80b6972b8fdd/Detectors/Base/include/DetectorsBase/CTFCoderBase.h#L312

Reply:
Thanks for ccdb explanations.
I modified CTFCoder and left CTFCoderBase without change.